scalafix-core/shared/src/main/scala/scalafix/internal/rule/DottyAutoTuplingFunctionArgs.scala

```diff
@@ -0,0 +1,68 @@
+package scalafix.internal.rule
+
+import scala.meta._
+import scala.meta.tokens.Token._
+import scalafix._
+import scalafix.syntax._
+import scalafix.util.SymbolMatcher
+
+case class DottyAutoTuplingFunctionArgs(index: SemanticdbIndex)
+    extends SemanticRule(index, "DottyAutoTuplingFunctionArgs") {
+  override def description: String =
+    "Rewrite that removes pattern-matching decomposition if function arguments can be automatically tupled in Dotty."
+
+  override def fix(ctx: RuleCtx): Patch =
+    ctx.tree.collect {
+      case Term.Apply((_, List(pf: Term.PartialFunction)))
+          if canBeAutoTupled(pf) =>
+        val pfTokens = pf.tokens
+
+        (for {
+          leftBrace <- pfTokens.headOption if leftBrace.is[LeftBrace]
+          rightBrace <- pfTokens.lastOption if rightBrace.is[RightBrace]
+          kwCase <- pfTokens.find(_.is[KwCase])
+        } yield {
+          val tkList = ctx.tokenList
+          val replacedBraces =
+            if (canReplaceBracesWithParens(pf)) {
+              ctx.replaceToken(leftBrace, "(") +
+                ctx.replaceToken(rightBrace, ")") +
+                ctx.removeTokens(tkList.leadingSpaces(leftBrace)) +
+                ctx.removeTokens(tkList.trailingSpaces(leftBrace)) +
+                ctx.removeTokens(tkList.leadingSpaces(rightBrace))
+            } else Patch.empty
+
+          replacedBraces +
+            ctx.removeToken(kwCase) +
+            ctx.removeTokens(tkList.trailingSpaces(kwCase))
+        }).asPatch
+    }.asPatch
+
+  private val ScalaAny = SymbolMatcher.exact(Symbol("_root_.scala.Any#"))
+  private def isMaybeAny(name: Term.Name): Boolean =
+    (for {
+      symbol <- name.symbol
+      tpe <- symbol.resultType
+    } yield ScalaAny.matches(tpe)).getOrElse(true)
+
+  // See https://github.com/scalacenter/scalafix/issues/521 why we
+  // skip the rewrite in this case.
+  private def allNamesAreAny(args: List[Pat]): Boolean = args.forall {
+    case Pat.Var(name) => isMaybeAny(name)
+    case _ => true
+  }
+
+  private def canBeAutoTupled(pf: Term.PartialFunction): Boolean =
+    pf.cases match {
+      case Case(Pat.Tuple(args), None, _) :: Nil =>
+        args.forall {
+          case Pat.Var(_) => true
+          case Pat.Wildcard() => true
+          case _ => false
+        } && !allNamesAreAny(args)
+      case _ => false
+    }
```

Comment:
there is one scenario that was pointed out here https://github.com/scalacenter/scalafix/pull/513#issuecomment-353062915 that should be skipped:

```scala
  def foo2(f: PartialFunction[(Int, Int), Int]): Int = ???
  foo2 { case (x, y) => 1 }
```

this should do the job:
```scala
  override def fix(ctx: RuleCtx): Patch = {
    ctx.tree.collect {
      case Term.Apply((Term.Select(_, name), List(pf: Term.PartialFunction)))
    // omitted
  }

  private def canBeAutoTupled(method: Term.Name, pf: Term.PartialFunction): Boolean =
    if (!hasPartialFunctionParam(method)) {
      // omitted
    } else false

  private val PartialFunction = SymbolMatcher.exact(Symbol("_root_.scala.PartialFunction#"))
  private def hasPartialFunctionParam(methodName: Term.Name): Boolean =
    (for {
      denotation <- index.denotation(methodName)
      paramTypeName <- denotation.names.headOption
      paramTypeSymbol = paramTypeName.symbol
    } yield PartialFunction.matches(paramTypeSymbol)).getOrElse(false)
```

Reply:
If I understand correctly, auto-tupling currently works fine for partial functions. In the dotty repl
```scala
@ def foo2(f: PartialFunction[(Int, Int), Int]): Int = f(1 -> 2)
defined function foo2

@   foo2 { case (x, y) => 1 }
res3: Int = 1
```

Has there been any change @allanrenucci ?

Reply:
That repl session was actually ammonite, but dotty is the same
```scala
scala> def foo2(f: PartialFunction[(Int, Int), Int]): Int = f(1 -> 2)
def foo2(f: PartialFunction[(Int, Int), Int]): Int
scala> foo2 { case (x, y) => 1 }
val res0: Int = 1
scala> foo2((x, y) => 1)
val res1: Int = 1
```

Reply:
auto-tupling currently works for any subtype of `FunctionN` and SAM types. There is an open issue that hasn't been discussed: lampepfl/dotty#3682